aten/src/ATen/native/cuda/group_norm_kernel.cu

```diff
@@ -28,80 +28,32 @@ constexpr int kCUDANumThreads = 256;
 constexpr int kReduceTileSize = 32;
 
 template <typename T>
-__global__ void RowwiseMomentsCUDAKernelNHWC(
-    int64_t N,
-    int64_t H,
-    int64_t W,
+__global__ void RowwiseMomentsCUDAKernel(
     int64_t C,
+    int64_t HxW,
     int64_t G,
     T eps,
     const T* X,
     T* mean,
-    T* rstd) {
+    T* rstd,
+    bool use_nchw) {
   using T_ACC = acc_type<T, true>;
   using WelfordType = WelfordData<T_ACC, int64_t>;
   using WelfordOp =
       WelfordOps<T_ACC, T_ACC, int64_t, thrust::pair<T_ACC, T_ACC>>;
 
-  const int64_t channels_per_group = C / G;
-  const int64_t batch_index = blockIdx.x / G;
-  const int64_t ng = blockIdx.x % G;
-  const int64_t batch_offset = batch_index * H * W * C;
-  const int64_t group_offset = ng * channels_per_group;
-  const int64_t start = batch_offset + group_offset;
-
   WelfordOp welford_op = {/*correction=*/0, /*take_sqrt=*/false};
   WelfordType val(0, 0, 0, 0);
-  for (int64_t j = threadIdx.x; j < H * W; j += blockDim.x) {
-    for (int64_t c = 0; c < channels_per_group; ++c) {
-      const int64_t index = start + j * C + c;
+  const int64_t n = blockIdx.x / G;
+  const int64_t g = blockIdx.x % G;
+  const int64_t D = C / G;
+  int batch_offset = n * C * HxW;
+  for (int64_t d = 0; d < D; ++d) {
+    for (int64_t hw = threadIdx.x; hw < HxW; hw += blockDim.x) {
+      const int64_t index = batch_offset + (use_nchw ? (g * D + d) * HxW + hw : hw * C + g * D + d);
```

Comment:
This is the part where I think it might be interesting to consider performance implications of access patterns:
In the NCHW case the accesses between neighboring threads (`threadIdx.x` and `threadIdx.x + 1`) appear to have stride 1 as `hw` is the offset. In NHWC the accesses have stride `C` between threads due to `hw * C` which could prevent memory coalescing. I'm thinking it might be possible to try rearrange the reduction so that neighboring threads have stride 1 once again.

Reply:
I agree although I initially didn't want to mess with memory coalescing because it would make the indexing logic more complicated. But I can work on a kernel that better handles this.

Reply:
Update: implemented memory coalesced NHWC kernels for fwd + bwd and NHWC now performs ~70-110% as fast as NCHW with the relatively slow inputs not being related to pixel count:

**FP32**
| Kernel           |   B (batch) |   C (num channels) |   R (resolution) |   G (num groups) |   Bandwidth (GB/s) |   Bandwidth Fraction of Peak |   Speedup over nn.GroupNorm NCHW |
|:-----------------|------------:|-------------------:|-----------------:|-----------------:|-------------------:|-----------------------------:|---------------------------------:|
| torch.nn GN NCHW |           1 |                 32 |               64 |               32 |              18.16 |                         0.06 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |               64 |               32 |              13.35 |                         0.05 |                             0.74 |
| GN NHWC          |           1 |                 32 |               64 |               32 |               5.06 |                         0.02 |                             0.28 |
| torch.nn GN NCHW |           1 |                 32 |              512 |               32 |             222.45 |                         0.77 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |              512 |               32 |             238.98 |                         0.83 |                             1.07 |
| GN NHWC          |           1 |                 32 |              512 |               32 |             236.92 |                         0.82 |                             1.07 |
| torch.nn GN NCHW |           1 |                128 |               64 |               32 |              56.15 |                         0.19 |                             1    |
| torch.nn GN NHWC |           1 |                128 |               64 |               32 |              49.51 |                         0.17 |                             0.88 |
| GN NHWC          |           1 |                128 |               64 |               32 |              36.78 |                         0.13 |                             0.66 |
| torch.nn GN NCHW |           1 |                128 |              512 |               32 |             237.85 |                         0.83 |                             1    |
| torch.nn GN NHWC |           1 |                128 |              512 |               32 |             261.56 |                         0.91 |                             1.1  |
| GN NHWC          |           1 |                128 |              512 |               32 |             261.39 |                         0.91 |                             1.1  |
| torch.nn GN NCHW |           8 |                 32 |               64 |               32 |             113.88 |                         0.4  |                             1    |
| torch.nn GN NHWC |           8 |                 32 |               64 |               32 |             100.89 |                         0.35 |                             0.89 |
| GN NHWC          |           8 |                 32 |               64 |               32 |              72.05 |                         0.25 |                             0.63 |
| torch.nn GN NCHW |           8 |                 32 |              512 |               32 |             266.28 |                         0.92 |                             1    |
| torch.nn GN NHWC |           8 |                 32 |              512 |               32 |             256.32 |                         0.89 |                             0.96 |
| GN NHWC          |           8 |                 32 |              512 |               32 |             263.12 |                         0.91 |                             0.99 |
| torch.nn GN NCHW |           8 |                128 |               64 |               32 |             229.94 |                         0.8  |                             1    |
| torch.nn GN NHWC |           8 |                128 |               64 |               32 |             221.72 |                         0.77 |                             0.96 |
| GN NHWC          |           8 |                128 |               64 |               32 |             217.23 |                         0.75 |                             0.94 |


**BF16** 
| Kernel           |   B (batch) |   C (num channels) |   R (resolution) |   G (num groups) |   Bandwidth (GB/s) |   Bandwidth Fraction of Peak |   Speedup over nn.GroupNorm NCHW |
|:-----------------|------------:|-------------------:|-----------------:|-----------------:|-------------------:|-----------------------------:|---------------------------------:|
| torch.nn GN NCHW |           1 |                 32 |               64 |               32 |               7.27 |                         0.03 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |               64 |               32 |               4.58 |                         0.02 |                             0.63 |
| GN NHWC          |           1 |                 32 |               64 |               32 |               2.39 |                         0.01 |                             0.33 |
| torch.nn GN NCHW |           1 |                 32 |              512 |               32 |             130.5  |                         0.45 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |              512 |               32 |             139.78 |                         0.49 |                             1.07 |
| GN NHWC          |           1 |                 32 |              512 |               32 |             130.81 |                         0.45 |                             1    |
| torch.nn GN NCHW |           1 |                128 |               64 |               32 |              27.25 |                         0.09 |                             1    |
| torch.nn GN NHWC |           1 |                128 |               64 |               32 |              20.16 |                         0.07 |                             0.74 |
| GN NHWC          |           1 |                128 |               64 |               32 |              11.23 |                         0.04 |                             0.41 |
| torch.nn GN NCHW |           1 |                128 |              512 |               32 |             176.08 |                         0.61 |                             1    |
| torch.nn GN NHWC |           1 |                128 |              512 |               32 |             197.96 |                         0.69 |                             1.12 |
| GN NHWC          |           1 |                128 |              512 |               32 |             246.93 |                         0.86 |                             1.4  |
| torch.nn GN NCHW |           8 |                 32 |               64 |               32 |              52.43 |                         0.18 |                             1    |
| torch.nn GN NHWC |           8 |                 32 |               64 |               32 |              39.37 |                         0.14 |                             0.75 |
| GN NHWC          |           8 |                 32 |               64 |               32 |              37.81 |                         0.13 |                             0.72 |
| torch.nn GN NCHW |           8 |                 32 |              512 |               32 |             212.23 |                         0.74 |                             1    |
| torch.nn GN NHWC |           8 |                 32 |              512 |               32 |             187.7  |                         0.65 |                             0.88 |
| GN NHWC          |           8 |                 32 |              512 |               32 |             232.94 |                         0.81 |                             1.1  |
| torch.nn GN NCHW |           8 |                128 |               64 |               32 |             174.75 |                         0.61 |                             1    |
| torch.nn GN NHWC |           8 |                128 |               64 |               32 |             159.06 |                         0.55 |                             0.91 |
| GN NHWC          |           8 |                128 |               64 |               32 |             183.8  |                         0.64 |                             1.05 |
| torch.nn GN NCHW |           8 |                128 |              512 |               32 |             204.28 |                         0.71 |                             1    |
| torch.nn GN NHWC |           8 |                128 |              512 |               32 |             184    |                         0.64 |                             0.9  |
| GN NHWC          |           8 |                128 |              512 |               32 |             260.68 |                         0.91 |                             1.28 |

Reply:
Update: implemented memory coalesced NHWC kernels for fwd + bwd and NHWC now performs ~70-110% as fast as NCHW with the relatively slow inputs not being related to pixel count:

**FP32**
| Kernel           |   B (batch) |   C (num channels) |   R (resolution) |   G (num groups) |   Bandwidth (GB/s) |   Bandwidth Fraction of Peak |   Speedup over nn.GroupNorm NCHW |
|:-----------------|------------:|-------------------:|-----------------:|-----------------:|-------------------:|-----------------------------:|---------------------------------:|
| torch.nn GN NCHW |           1 |                 32 |               64 |               32 |              18.16 |                         0.06 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |               64 |               32 |              13.35 |                         0.05 |                             0.74 |
| GN NHWC          |           1 |                 32 |               64 |               32 |               5.06 |                         0.02 |                             0.28 |
| torch.nn GN NCHW |           1 |                 32 |              512 |               32 |             222.45 |                         0.77 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |              512 |               32 |             238.98 |                         0.83 |                             1.07 |
| GN NHWC          |           1 |                 32 |              512 |               32 |             236.92 |                         0.82 |                             1.07 |
| torch.nn GN NCHW |           1 |                128 |               64 |               32 |              56.15 |                         0.19 |                             1    |
| torch.nn GN NHWC |           1 |                128 |               64 |               32 |              49.51 |                         0.17 |                             0.88 |
| GN NHWC          |           1 |                128 |               64 |               32 |              36.78 |                         0.13 |                             0.66 |
| torch.nn GN NCHW |           1 |                128 |              512 |               32 |             237.85 |                         0.83 |                             1    |
| torch.nn GN NHWC |           1 |                128 |              512 |               32 |             261.56 |                         0.91 |                             1.1  |
| GN NHWC          |           1 |                128 |              512 |               32 |             261.39 |                         0.91 |                             1.1  |
| torch.nn GN NCHW |           8 |                 32 |               64 |               32 |             113.88 |                         0.4  |                             1    |
| torch.nn GN NHWC |           8 |                 32 |               64 |               32 |             100.89 |                         0.35 |                             0.89 |
| GN NHWC          |           8 |                 32 |               64 |               32 |              72.05 |                         0.25 |                             0.63 |
| torch.nn GN NCHW |           8 |                 32 |              512 |               32 |             266.28 |                         0.92 |                             1    |
| torch.nn GN NHWC |           8 |                 32 |              512 |               32 |             256.32 |                         0.89 |                             0.96 |
| GN NHWC          |           8 |                 32 |              512 |               32 |             263.12 |                         0.91 |                             0.99 |
| torch.nn GN NCHW |           8 |                128 |               64 |               32 |             229.94 |                         0.8  |                             1    |
| torch.nn GN NHWC |           8 |                128 |               64 |               32 |             221.72 |                         0.77 |                             0.96 |
| GN NHWC          |           8 |                128 |               64 |               32 |             217.23 |                         0.75 |                             0.94 |


**BF16** 
| Kernel           |   B (batch) |   C (num channels) |   R (resolution) |   G (num groups) |   Bandwidth (GB/s) |   Bandwidth Fraction of Peak |   Speedup over nn.GroupNorm NCHW |
|:-----------------|------------:|-------------------:|-----------------:|-----------------:|-------------------:|-----------------------------:|---------------------------------:|
| torch.nn GN NCHW |           1 |                 32 |               64 |               32 |               7.27 |                         0.03 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |               64 |               32 |               4.58 |                         0.02 |                             0.63 |
| GN NHWC          |           1 |                 32 |               64 |               32 |               2.39 |                         0.01 |                             0.33 |
| torch.nn GN NCHW |           1 |                 32 |              512 |               32 |             130.5  |                         0.45 |                             1    |
| torch.nn GN NHWC |           1 |                 32 |              512 |               32 |             139.78 |                         0.49 |                             1.07 |
| GN NHWC          |           1 |                 32 |              512 |               32 |             130.81 |                         0.45 |                             1    |
| torch.nn GN NCHW |           1 |                128 |               64 |               32 |              27.25 |                         0.09 |                             1    |
| torch.nn GN NHWC |           1 |                128 |               64 |               32 |              20.16 |                         0.07 |                             0.74 |
| GN NHWC          |           1 |                128 |               64 |               32 |              11.23 |                         0.04 |                             0.41 |
| torch.nn GN NCHW |           1 |                128 |              512 |               32 |             176.08 |                         0.61 |                             1    |
| torch.nn GN NHWC |           1 |                128 |              512 |               32 |             197.96 |                         0.69 |                             1.12 |
| GN NHWC          |           1 |                128 |              512 |               32 |             246.93 |                         0.86 |                             1.4  |
| torch.nn GN NCHW |           8 |                 32 |               64 |               32 |              52.43 |                         0.18 |                             1    |
| torch.nn GN NHWC |           8 |                 32 |               64 |               32 |              39.37 |                         0.14 |                             0.75 |
| GN NHWC          |           8 |                 32 |               64 |               32 |              37.81 |                         0.13 |                             0.72 |
| torch.nn GN NCHW |           8 |                 32 |              512 |               32 |             212.23 |                         0.74 |                             1    |
| torch.nn GN NHWC |           8 |                 32 |              512 |               32 |             187.7  |                         0.65 |                             0.88 |
| GN NHWC          |           8 |                 32 |              512 |               32 |             232.94 |                         0.81 |                             1.1  |
| torch.nn GN NCHW |           8 |                128 |               64 |               32 |             174.75 |                         0.61 |                             1    |
| torch.nn GN NHWC |           8 |                128 |               64 |               32 |             159.06 |                         0.55 |                             0.91 |
| GN NHWC          |           8 |                128 |               64 |               32 |             183.8  |                         0.64 |                             1.05 |
| torch.nn GN NCHW |           8 |                128 |              512 |               32 |             204.28 |                         0.71 |                             1    |
| torch.nn GN NHWC |           8 |                128 |              512 |               32 |             184    |                         0.64 |                             0.9  |
| GN NHWC          |           8 |                128 |              512 |               32 |             260.68 |                         0.91 |                             1.28 |